### PR TITLE
feat(harness): add a project with the OS folder browser on desktop [SAP-3143]

### PR DESCRIPTION
**Pull request by gwitwer**
## Primary change type

- [x] Feature

## Problem and motivation

Adding a project on the desktop app opened a modal wrapping a text input, with Finder or Explorer offered as a button *inside* it. The desktop/web split already existed but sat one level too deep: `FolderField` feature-detects `getDesktopBridge()?.chooseDirectory` and uses the answer to decide whether to render a `<datalist>`, so the bridge changed what rendered inside our dialog rather than whether the dialog was the right control at all.

With a real OS folder browser available, our dialog is not a smaller version of it — it is a different and worse control. The typed path is what a browser host is left with when it has nothing better.

This is an acceptance item on SAP-3143: "The folder step opens the native picker when the desktop bridge is present; typed path only on web."

## Summary and scope

The rail's Add-a-project `+` now asks `chooseProjectFolder`, and the bridge decides whether our dialog opens at all. On desktop the OS browser opens directly and the chosen folder goes straight to `onOpenProject`. On web, behaviour is unchanged.

The decision lives in `web/src/lib/folder-step.ts` rather than inline at the caller, so hosts cannot drift apart as more entrances start asking this same question.

**Intentionally out of scope:**

- `StartDialog`'s `detect` mode ("Add existing agents") is untouched. It has to show what it found under the folder, so the picker is only its first step there; `open` mode asks one question with one action, which is why only it can be answered outright by the OS browser.
- The rest of SAP-3143 — the create verb's form, removing `composerScaffoldPrompt` (still live at `web/src/App.tsx:1888`), and making the composer project-aware. Those carry open design questions the ticket flags explicitly, including an unresolved question against SAP-2981 about whether any prompt is submitted at all. **SAP-3143 stays open.**

## Related work

Related issue or discussion: SAP-3143.

Salvaged from #803, which is now closed. That PR was stacked on `gwitwer/studio-rail-creation` (its own PR #801 closed unmerged, superseded by #819 + `[Agent Map 01..15/15]`), and its central commit dispatched to `harness.openPlannerSession` — deleted from `main` by `21684912` "feat(harness): activate ordinary bootstrap [Agent Map 07/15]", which is the founder's reversal of the planner session type. This file is the one piece of #803 with no open design question. See the closing comment on #803 for the full accounting of what was and was not carried forward.

## Validation

```text
npx vitest run web/src/lib/folder-step.test.ts    — 6 passed (6)
npx playwright test folder-field.spec.ts          — 6 passed (6)
npx playwright test open-project unrooted-agents \
  project-axis create-agent agent-map-navigation \
  agent-map-initialization smoke                  — 180 passed (180)
npx tsc --noEmit                                  — clean
npx tsc --noEmit -p web/tsconfig.json             — clean
npx eslint src --ext .ts                          — clean
npm run build (tsc + vite web bundle)             — built, no new warnings
npx prettier --check <changed files>              — clean on changed lines
```

`pnpm --filter "@sapiom/harness^..." build` is required before typecheck, or `@sapiom/agent-core` and `@sapiom/mcp` report as missing modules.

Note: `WorkflowsRail.tsx` already fails `prettier --check` on `main` at lines 712-720 and 754-757. Left untouched to keep this diff to one problem.

### Tests and documentation

`web/src/lib/folder-step.test.ts`, 6 cases, plus 4 desktop-host cases in `web/e2e/folder-field.spec.ts`.

The desktop branch is unreachable by both test tiers — no Electron in the unit environment, and Playwright is a browser host by definition — so `chooseProjectFolder` is a pure function over an injected host, and the spec asserts the negatives: that our dialog does **not** also open when the bridge is present, and that a **cancelled** pick does not fall back into it. Dismissing a folder browser is the user declining the question, not declining the way it was asked.

**A failed pick is a different answer, and it does fall back.** Cancelling resolves `null`; only a broken bridge rejects, and swallowing that left Add a project doing nothing at all on a host whose picker was unavailable. Both sides are pinned so they cannot re-converge: the e2e cancel case asserts no dialog opens, the unit failure case asserts one does.

Two gaps, stated rather than papered over: the starting-folder tier selection is a JSX expression with no direct test (the helper's empty-value handling is covered for both `null` and `""`, but "which of the three tiers wins" is not), and `StartDialog.tsx:82` has the same `??`-versus-empty-string gap for those tiers — left alone because it is reachable from other entrances and is not this PR's change.

No user-facing docs cover this control; the behaviour change is in the changeset.

## Compatibility and release impact

- Breaking or externally visible changes: None breaking. Externally visible: on the desktop app, Add a project opens the OS folder browser instead of a typed-path dialog. Web is unchanged.
- Changeset: Added — `.changeset/native-folder-step.md` (patch).

## Security

- [x] I have not included secrets, credentials, private data, or unsanitized logs.
- [x] This pull request does not publicly disclose a suspected vulnerability.

## AI assistance

- [x] I used AI assistance and have described it below.

Claude Code investigated #803's viability, established that its base was abandoned and its core mechanism deleted from main, and selected and adapted this file from that branch. It wrote the caller wiring and the doc comments, and ran the build, typecheck, lint, and test commands quoted above. `folder-step.ts` and its spec are substantially the originals from #803 with the doc comments corrected to describe this narrower change rather than the create-verb redesign, which is not in this PR.

## Checklist

- [x] I read `CONTRIBUTING.md`, and this contribution follows the direct-PR or issue-first policy.
- [x] This pull request addresses one focused problem and contains no unrelated cleanup.
- [x] I added or updated tests, or explained above why tests are not applicable.
- [x] I ran the relevant build, typecheck, lint, and test commands, or explained any N/A checks above.
- [x] I updated documentation for user-facing changes, or marked it N/A above.
- [x] I added a Changeset for a published-package change, or explained why it is not applicable.
- [x] I can explain and maintain every submitted change, including any AI-assisted work.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_012wSFNBtYuYRA2XQunaVMvV

## Summary by CodeRabbit

* **New Features**
  * Desktop users can open the operating system’s folder picker directly when creating a project.
  * Selected folders open automatically without an additional confirmation step.
  * If the native picker is unavailable, project creation continues to use the existing web dialog.
  * “Add existing agents” continues to use the in-app dialog.

* **Bug Fixes**
  * Cancelled or unsuccessful folder selections leave the current view unchanged.
  * Project selections remain correct when folder scanning and user actions overlap.

---

<a href="https://app.devin.ai/review/sapiom/sapiom-js/pull/915" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-devin-review-dark.svg?v=3">
    Devin Review
  </picture>
</a>

### CODE REVIEW

File: packages/harness/web/src/lib/folder-step.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * dismissing a folder browser is not a failure, and there is nothing to report.
+ */
+export async function chooseProjectFolder(host: FolderStepHost): Promise<void> {
+  if (!host.chooseDirectory) {
+    host.openDialog();
+    return;
+  }
+  let picked: string | null = null;
+  try {
+    picked = await host.chooseDirectory(host.startingAt ?? undefined);
```
[the remark this answers is not in the document]
Comment (gwitwer) [reply packages/harness/web/src/lib/folder-step.ts]: Taken, in `946ab3dd` — and it was reachable rather than theoretical, which is worth recording.

`App.tsx` normalizes one of the two inputs and not the other: `projectRoot={projectRoot || null}` (`:2840`) drops `""`, but `launchDir={state.launchDir ?? null}` (`:2972`) does not. The call site passes `projectRoot ?? launchDir`, so an empty `launchDir` from the server arrives here as `""` and `?? undefined` forwards it as an empty `defaultPath`.

I used `host.startingAt || undefined` rather than the suggested ternary, since `||` covers `null`, `undefined` and `""` in one expression and this helper has no meaningful use for an empty path.

Also note the existing test named "omits the starting folder rather than passing an empty one" only ever passed `null` — the case its own name describes was untested. Added as a separate case (now 6) rather than changing that one, so both inputs stay covered.

**Review by ynadge — CHANGES_REQUESTED** [review]
Small UX change: A slow folder open can override a newer selection. After choosing folder A, the native picker closes and the rail remains usable while A is scanned. If you select existing project B, finishing A’s scan switches the view back to A.

**Comment by gwitwer** [comment]
Good catch, and it was worse than a UX nit — thanks. Fixed in `52cb4ec6`.

**Why it happened.** The dialog used to be the lock. While it was open the rail sat behind a modal, so "add a folder" could not overlap "select a project". Opening the OS picker hands the rail back the moment it closes, and the scan behind it is the slow part — so the two overlap now, exactly as you describe.

**The guard already existed and was claimed too late.** `handleSelectWorkspace` bumps `studioRestoreGenerationRef` (`App.tsx:1696`), so selecting B *did* invalidate the pending open. But `onOpenProject` took its generation **after** `openProject` and `getState`, claiming a newer number than the selection you made while waiting, then proceeding to navigate. It won a race it had already lost. It now claims before the scan and re-checks after each await.

The open itself stays unconditional — the project is still added, since you did ask for it. Only the navigation is dropped. I gated the `catch` fallback the same way: failing to read a preference is not a reason to override a later choice. Its generation bump moved inside that gate too, or a stale failure would invalidate the newer selection's own pending restore.

**Scope note:** this entrance exposed it, but the handler is shared — "Open as project" on an unrooted agent and the web dialog had the same race already, with no modal to hide it. Those are fixed as a consequence. Say the word if you'd rather I narrow it to only the new path, though I think the shared fix is the right call.

**On the test, since it nearly fooled me.** My first version passed against the *unfixed* build. `toHaveAttribute` retries until it matches and returns on the first pass, so asserting as soon as the new project row appears is satisfied by the state *before* the stale navigation lands. It now settles first and asserts from both sides — B still pressed, and the late arrival did not take the selection. Confirmed failing without the fix (`Expected: "true" Received: "false"`) and passing with it.

Verified: `folder-field` 6/6; `open-project`, `unrooted-agents`, `project-axis`, `create-agent`, `agent-map-navigation`, `agent-map-initialization`, `smoke` 180/180; typecheck and build clean.

File: packages/harness/web/src/lib/folder-step.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    return;
+  }
+  let picked: string | null = null;
+  try {
+    // `||`, NOT `??`: an empty string has to be omitted too, not forwarded as
+    // an empty `defaultPath`. `launchDir` reaches callers through `?? null`,
+    // which lets `""` past, and "start at nowhere" is not a starting folder.
+    picked = await host.chooseDirectory(host.startingAt || undefined);
+  } catch {
+    return;
```
[the remark this answers is not in the document]
Comment (gwitwer) [reply packages/harness/web/src/lib/folder-step.ts:72]: Correct, taken in `a4898703`.

The bug behind it is that cancelling and failing were conflated. Cancelling resolves `null`; only a broken bridge rejects. The doc comment justified the swallow with reasoning that only covers cancel — "dismissing a folder browser is not a failure" — and the failure path inherited it. The consequence is exactly what you describe: on a desktop build whose picker is unavailable, the one entrance to adding a project does nothing, with no feedback and no way through.

A rejection now opens the typed-path dialog, which still works on that host. Cancelling is unchanged. Both sides are pinned so they cannot re-converge: the e2e cancel case asserts no dialog opens, and the unit failure case asserts one does. I rewrote the unit test rather than adding one, since it was named "swallows a failed native pick without falling back" — it asserted the old behaviour by name.

File: packages/harness/web/src/components/WorkflowsRail.tsx [hunk 3]
```diff
@@ -1006,7 +1007,28 @@ export function WorkflowsRail({
...
+              // renders inside it. `FolderField` already feature-detects
+              // `chooseDirectory`, but one level too deep: it used the answer
+              // to swap a `<datalist>`, so a desktop user still got a modal
+              // wrapping a text input with the OS browser offered as a button
+              // inside it. In `open` mode this dialog asks one question — which
+              // folder — and Finder answers it better than we do.
+              void chooseProjectFolder({
+                chooseDirectory: getDesktopBridge()?.chooseDirectory ?? null,
+                // The dialog's own precedence for where to start.
+                startingAt: projectRoot ?? launchDir,
```
[the remark this answers is not in the document]
Comment (gwitwer) [reply packages/harness/web/src/components/WorkflowsRail.tsx]: Correct, taken in `a4898703`.

Verified against the dialog this replaces: `StartDialog.tsx:82` starts at `projectRoot ?? launchDir ?? recentDirs[0] ?? ""`. I carried only the first two tiers, so the case the third exists for — no project open, nothing pinned, but somewhere the user was working recently — sent the OS picker to its own default folder instead.

Now `projectRoot ?? launchDir ?? recentDirs[0]`, matching the dialog exactly. Empty values still normalize away in `chooseProjectFolder` via `|| undefined`, so an empty `recentDirs[0]` cannot become an empty `defaultPath` either.

File: packages/harness/web/src/components/WorkflowsRail.tsx [hunk 4]
```diff
@@ -1006,7 +1007,32 @@ export function WorkflowsRail({
...
+              // inside it. In `open` mode this dialog asks one question — which
+              // folder — and Finder answers it better than we do.
+              void chooseProjectFolder({
+                chooseDirectory: getDesktopBridge()?.chooseDirectory ?? null,
+                // The dialog's own precedence for where to start, all three
+                // tiers of it (`StartDialog`'s initial `cwd`). Dropping the
+                // last one sent the OS picker to its default folder in the
+                // case the dialog handled best: no project open, nothing
+                // pinned, but somewhere the user was working recently.
+                startingAt: projectRoot ?? launchDir ?? recentDirs[0],
```
[the remark this answers is not in the document]
Comment (gwitwer) [reply packages/harness/web/src/components/WorkflowsRail.tsx]: Correct, taken in `91fda29a`. Same class of bug as the first review finding on this PR — which I described in the commit message two commits ago and then wrote into the tier chain myself, so thanks for catching it.

`??` stops at an empty `launchDir` because `""` is neither null nor undefined, never reaches `recentDirs[0]`, and the helper then discards the empty value — OS default, with a usable recent folder sitting right there.

Now `projectRoot || launchDir || recentDirs[0]`. Worth flagging that this is a deliberate divergence from `StartDialog.tsx:82`, which writes `??` for the same three tiers: an empty string is not an answer to "which folder", so a tier should fall through when empty, not only when absent. The dialog has the same latent gap. I have not touched it, since it is not this PR's change and it is reachable from other entrances — say the word if you would rather it were fixed here or tracked separately.

One honest gap: the tier selection itself is a JSX expression with no direct test. The helper's empty-value handling is covered (`folder-step.test.ts`, both `null` and `""`), but "which of the three tiers wins" is not. Extracting a three-line expression to test it felt worse than saying so.

File: packages/harness/web/src/lib/folder-step.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  let picked: string | null = null;
+  try {
+    // `||`, NOT `??`: an empty string has to be omitted too, not forwarded as
+    // an empty `defaultPath`. `launchDir` reaches callers through `?? null`,
+    // which lets `""` past, and "start at nowhere" is not a starting folder.
+    picked = await host.chooseDirectory(host.startingAt || undefined);
+  } catch {
+    // Not a decline — see above. Leave the user a working way to answer.
+    host.openDialog();
+    return;
```
[the remark this answers is not in the document]
Comment (gwitwer) [reply packages/harness/web/src/lib/folder-step.ts:72]: Right, and the description was the stale half — the code is what I meant it to be. Updated.

The body was written before `a4898703`, so it still carried the sentence "a cancelled or failed native pick does not fall back into it", which described the behaviour your own earlier finding got me to change. It now says the two are different answers: cancelling resolves `null` and does nothing; a rejection means the bridge is broken and falls back to the typed-path dialog.

While in there I refreshed the rest of the description against the current head — test counts (6 unit, 6 e2e, 180 across the neighbouring specs) and the two coverage gaps I would rather have written down than left for a reviewer to discover: the tier-selection expression has no direct test, and `StartDialog.tsx:82` still has the same `??`-versus-empty-string gap, deliberately untouched here.

**Comment by gwitwer** [comment]
Closing: superseded by #1075 (SAP-3573), which moves the same `chooseProjectFolder` helper and the Add-project entrance into App.tsx and also removes StartDialog from the rail. The one human review request here (a slow folder scan overriding a newer selection) was fixed in 52cb4ec6; that fix and two smaller ones from this PR's review rounds are not in #1075 yet, noted there.
